cmd/kubeadm/app/cmd/upgrade/common.go

```diff
@@ -24,127 +24,99 @@ import (
-
-// enforceRequirements verifies that it's okay to upgrade and then returns the variables needed for the rest of the procedure
-func enforceRequirements(flags *applyPlanFlags, args []string, dryRun bool, upgradeApply bool, printer output.Printer, loadConfig LoadConfigFunc) (clientset.Interface, upgrade.VersionGetter, *kubeadmapi.InitConfiguration, error) {
-	client, err := getClient(flags.kubeConfigPath, dryRun)
-	if err != nil {
-		return nil, nil, nil, errors.Wrapf(err, "couldn't create a Kubernetes client from file %q", flags.kubeConfigPath)
+	// Ensure the user is root
+	klog.V(1).Info("running preflight checks")
+	if err := runPreflightChecks(client, ignorePreflightErrorsSet, printer); err != nil {
+		return nil, nil, nil, nil, err
 	}
 
-	// Fetch the configuration from a file or ConfigMap and validate it
-	_, _ = printer.Println("[upgrade/config] Loading the kubeadm configuration")
-
-	var newK8sVersion string
-	cfg, legacyReconfigure, err := loadConfig(flags.cfgPath, client, !upgradeApply, printer)
+	initCfg, err := configutil.FetchInitConfigurationFromCluster(client, printer, "upgrade/config", false, true)
 	if err != nil {
-		return nil, nil, nil, errors.Wrap(err, "could not load the kubeadm configuration")
-	} else if legacyReconfigure {
-		// Set the newK8sVersion to the value in the ClusterConfiguration. This is done, so that users who use the --config option
-		// to supply a new ClusterConfiguration don't have to specify the Kubernetes version twice,
-		// if they don't want to upgrade but just change a setting.
-		newK8sVersion = cfg.KubernetesVersion
+		if apierrors.IsNotFound(err) {
+			_, _ = printer.Printf("[upgrade/config] In order to upgrade, a ConfigMap called %q in the %s namespace must exist.\n", constants.KubeadmConfigConfigMap, metav1.NamespaceSystem)
+			_, _ = printer.Printf("[upgrade/config] Next steps:\n")
+			_, _ = printer.Printf("\t- OPTION 1: Run 'kubeadm config upload from-flags' and specify the same CLI arguments you passed to 'kubeadm init' when you created your control-plane.\n")
+			_, _ = printer.Printf("\t- OPTION 2: Run 'kubeadm config upload from-file' and specify the same config file you passed to 'kubeadm init' when you created your control-plane.\n")
+			_, _ = printer.Println()
+			err = errors.Errorf("the ConfigMap %q in the %s namespace used for getting configuration information was not found", constants.KubeadmConfigConfigMap, metav1.NamespaceSystem)
+		}
+		return nil, nil, nil, nil, errors.Wrap(err, "[upgrade/init config] FATAL")
 	}
 
-	// The version arg is mandatory, during upgrade apply, unless it's specified in the config file
-	if upgradeApply && newK8sVersion == "" {
-		if err := cmdutil.ValidateExactArgNumber(args, []string{"version"}); err != nil {
-			return nil, nil, nil, err
+	newK8sVersion := upgradeCfg.Plan.KubernetesVersion
```

Comment:
for `kubeadm plan` and `kubeadm apply` command, we can first use  option `kubernetesVersion` and override it with version command parameter.